File: common/src/main/java/com/cc/dc/common/ui/BaseActivity.java
package com.cc.dc.common.ui;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.cc.dc.common.presenter.BaseModel;
import com.cc.dc.common.presenter.BasePresenter;

import butterknife.ButterKnife;

/**
 * Created by dc on 2017/9/18.
 */
public abstract class BaseActivity<T extends BasePresenter, E extends BaseModel> extends AppCompatActivity  {

    private T presenter;

    private E model;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getLayoutId() == 0) {
            throw new IllegalArgumentException("没有设置Layout");
        }
        setContentView(getLayoutId());
        ButterKnife.bind(this);
        initView();
        initPresenter();
//        EventBus.getDefault().register(this);
    }

    public abstract int getLayoutId();

    public abstract void initPresenter();

    public abstract void initView();

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onRestart() {
        super.onRestart();
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    @Override
    protected void onDestroy() {
//        EventBus.getDefault().unregister(this);
        super.onDestroy();
    }
}

File: kotlindemo/src/main/java/com/cc/dc/kotlindemo/base/BaseActivity.kt
package com.cc.dc.kotlindemo.base

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import butterknife.ButterKnife
import de.greenrobot.event.EventBus


/**
 * Created by dc on 2018/5/29.
 * BaseActivity
 */
abstract class BaseActivity : AppCompatActivity() {

    protected val pageSize = 15

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayout())
        if (isRegisterEventBus()) {
            EventBus.getDefault().register(this)
        }
        ButterKnife.bind(this)
        initView()
    }

    protected abstract fun getLayout() : Int

    protected open abstract fun initView()

    protected open fun isRegisterEventBus() : Boolean {
        return false
    }

    override fun onDestroy() {
        if (isRegisterEventBus()) {
            EventBus.getDefault().unregister(this)
        }
        super.onDestroy()
    }
}